[5.7] Added test helper to assert that a Job has been queued with a Chain

asantibanez opened the pull request:
The following PR adds a helper method for testing that allows to assert if a Job was pushed to the Queue with a specific Chain. The proposed method name is ```assertPushedWithChain``` and expects two parameters: the First Job in the Chain and an array of Jobs that will execute. 

I have attached an image on how this assertion can be used.

![image](https://user-images.githubusercontent.com/5126648/36941923-a5a590b4-1f34-11e8-90ac-b197167ba5a4.png)

I have used the ```Queueable``` trait ```chain``` method to simulate jobs serialization and have used the same method to check for the chain in the assertion method.

![image](https://user-images.githubusercontent.com/5126648/36941940-f9375e1a-1f34-11e8-9f1a-8221e73f4aa9.png)

This helper can be useful to assert that a chain has been triggered with a specific set of jobs and order.

Hope it is useful for others! Cheers!

asantibanez committed:
Added test helper to assert that a Job has been queued with a Chain

asantibanez committed:
Removed previous implementation dead code

asantibanez commented:
I am sorry about the second commit for this PR. I removed some dead code I had from my previous implementation for the same feature.

asantibanez commented:
I made some extra digging and found out that when the expected chain assertion fails, it gives a not so helpful message when checking the compared values (at least on PhpStorm) 

![image](https://user-images.githubusercontent.com/5126648/36942231-56dd8c1e-1f3b-11e8-9fb6-6baeec9fa1d0.png)

However, working with serialization the other way around, the compared values are more meaningful for debugging.

![image](https://user-images.githubusercontent.com/5126648/36942240-8ef1e726-1f3b-11e8-901b-f0f856cd35de.png)

This is done by unserializing the jobs chain instead of serializing the expected chain.

![image](https://user-images.githubusercontent.com/5126648/36942251-dd66ac20-1f3b-11e8-876a-6ca7192be4d6.png)

This could be decided going forward if the PR is a welcomed addition. 

Thanks.

taylorotwell commented:
What if you pushed the same job once without a chain and once with chain. Wouldn't this fail since you are only checking the first pushed job for the chain?

asantibanez commented:
You are right @taylorotwell . Never crossed my mind on testing if the same job was queued but with different payload. Let me check on how to deal with this scenario. I imagine it has to be similar than using a call back to check if the correct job was queued.

Is the proposed feature using correctly the job serialization/unserialization? I was also thinking on just testing for the class name to be chained instead of the object. What are your thoughts on this one?

Thanks.

asantibanez committed:
Added support to check for chained jobs using class names

- Added callback to support identifying particular jobs
- Added tests

asantibanez commented:
@taylorotwell 

I have added support for checking for particular jobs using a callback. This will cover the scenario you spoke of:

> What if you pushed the same job once without a chain and once with chain. Wouldn't this fail since you are only checking the first pushed job for the chain?

I have also added support for checking for chained jobs using class names. Now it is possible to check using an array of job class names or an array of job objects

![image](https://user-images.githubusercontent.com/5126648/37217167-cc181e74-238a-11e8-9695-e55c5eb4d03d.png)

Hope it is all good. Thanks for the feedback.

Regards.

asantibanez committed:
Removed unused dependencies

taylorotwell commented:
I still see you taking the first job. What if I push the same job twice with different chains? I want to assert that the job was pushed with each chain, so I would call assertPushedWithChain twice for the same job type but different chains. I would expect them both to pass. Feel free to re-visit when that has been solved.
